llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp

```diff
@@ -2092,6 +2100,14 @@ bool SIInsertWaitcnts::generateWaitcntInstBefore(MachineInstr &MI,
   // Verify that the wait is actually needed.
   ScoreBrackets.simplifyWaitcnt(Wait);
 
+  // Since the translation for VMEM addresses occur in-order, we can apply the
+  // XCnt if the current instruction is of VMEM type and has a memory
+  // dependency with another VMEM instruction in flight.
+  if (Wait.XCnt != ~0u && isVmemAccess(MI)) {
+    ScoreBrackets.applyWaitcnt(X_CNT, Wait.XCnt);
+    Wait.XCnt = ~0u;
+  }
+
```

Comment:
This was a bit of a misleading comment, unless this is a bug a vmemaccess instruction _was_ actually applying the xcnt, since the generateWaitcnt function was first applying it, since it was about to generate the waitcnt, but then before that could happen the previous implementation set it to ~0u. Ready for retrigger 😃 @jayfoad